services/data/postgres_async_db.py

```diff
@@ -668,33 +668,71 @@ class AsyncTaskTablePostgres(AsyncPostgresTable):
     keys = ["flow_id", "run_number", "run_id", "step_name", "task_id",
             "task_name", "user_name", "ts_epoch", "last_heartbeat_ts", "tags", "system_tags"]
     primary_keys = ["flow_id", "run_number", "step_name", "task_id"]
-    joins = ["LEFT JOIN {artifacts_table} AS artifacts ON ({table_name}.flow_id = artifacts.flow_id AND {table_name}.task_id = artifacts.task_id AND artifacts.name = '_task_ok')"
-             .format(table_name=table_name, artifacts_table="artifact_v3")]
+    # NOTE: There is a lot of unfortunate backwards compatibility support in the following join, due to
+    # the older metadata service not recording separate metadata for task attempts. This is also the
+    # reason why we must join through the artifacts table, instead of directly from metadata.
+    joins = [
+        """
+        LEFT JOIN (
+            SELECT
+                task_ok.flow_id, task_ok.run_number, task_ok.step_name,
+                task_ok.task_id, task_ok.attempt_id, task_ok.ts_epoch,
+                attempt.ts_epoch as started_at,
+                COALESCE(done.ts_epoch, task_ok.ts_epoch) as finished_at
+            FROM {artifact_table} as task_ok
+            LEFT JOIN {metadata_table} as attempt ON (
+                task_ok.flow_id = attempt.flow_id AND
+                task_ok.step_name = attempt.step_name AND
+                task_ok.task_id = attempt.task_id AND
+                attempt.field_name = 'attempt' AND
+                task_ok.attempt_id = attempt.value::int
+            )
+            LEFT JOIN {metadata_table} as done ON (
+                task_ok.flow_id = done.flow_id AND
+                task_ok.step_name = done.step_name AND
+                task_ok.task_id = done.task_id AND
+                done.field_name = 'attempt-done' AND
+                task_ok.attempt_id = done.value::int
+            )
```

Comment:
if absolute accuracy isn't a concern in favor of performance, this join could be omitted, and we can simply use the task_ok.ts_epoch as the finished_at in every case.